src/backend/storage/ipc/temp_tables_limit.c

```diff
@@ -0,0 +1,268 @@
+#include <libgen.h>
+#include <sys/stat.h>
+
+#include "postgres.h"
+
+#include "c.h"
+#include "cdb/cdbbufferedappend.h"
+#include "cdb/cdbvars.h"
+#include "pg_config_manual.h"
+#include "port/atomics.h"
+#include "storage/fd.h"
+#include "storage/relfilenode.h"
+#include "storage/shmem.h"
+#include "storage/smgr.h"
+#include "storage/temp_tables_limit.h"
+#include "utils/elog.h"
+#include "utils/guc.h"
+
+#define TempTablesLimitChecks() do {\
+									if (!IsUnderPostmaster || Gp_role != GP_ROLE_EXECUTE)\
+										return;\
+									\
+									Assert(temp_tables_limit_value != NULL);\
+								} while (0)
+
+static volatile pg_atomic_uint64 *temp_tables_limit_value = NULL;
+static int64 prevFileLen = -1;
+static int64 prevSegFileLen = -1;
+static bool fileSkip = false;
+static bool segFileSkip = false;
+
+static int64
+TempTablesLimitToBytes(void)
+{
+	return temp_tables_limit * 1024 * 1024;
+}
+
+/* init */
+
+Size
+TempTablesLimitSize(void)
+{
+	return sizeof(pg_atomic_uint64);
+}
+
+void
+TempTablesLimitShmemInit(void)
+{
+	temp_tables_limit_value = (volatile pg_atomic_uint64 *) ShmemAlloc(sizeof(pg_atomic_uint64));
+	pg_atomic_init_u64((pg_atomic_uint64 *) temp_tables_limit_value, 0);
+}
+
+/* ao */
+
+void
+BufferedAppendWritePreHook(BufferedAppend *bufferedAppend)
+{
+	TempTablesLimitChecks();
+
+	if (RelFileNodeBackendIsTemp(bufferedAppend->relFileNode))
+	{
+		prevFileLen = bufferedAppend->largeWritePosition;
+
+		if (bufferedAppend->largeWriteLen > 0 &&
+			temp_tables_limit_value->value + bufferedAppend->largeWriteLen > TempTablesLimitToBytes())
+			elog(ERROR, "Temp tables quota exceeded");
+	}
+}
+
+void
+BufferedAppendWritePostHook(BufferedAppend *bufferedAppend)
+{
+	TempTablesLimitChecks();
+
+	if (RelFileNodeBackendIsTemp(bufferedAppend->relFileNode))
+	{
+		Assert(prevFileLen >= 0);
+		pg_atomic_add_fetch_u64(temp_tables_limit_value, bufferedAppend->largeWriteLen);
+	}
+
+	prevFileLen = -1;
+}
+
+void
+TruncateAOSegmentFilePreHook(Relation rel, File fd)
+{
+	TempTablesLimitChecks();
+
+	if (rel->rd_islocaltemp)
+	{
+		prevFileLen = FileDiskSize(fd);
+	}
+}
+
+void
+TruncateAOSegmentFilePostHook(Relation rel, int64 offset)
+{
+	int64 bytesToTruncate;
+
+	TempTablesLimitChecks();
+
+	if (rel->rd_islocaltemp)
+	{
+		Assert(prevFileLen >= 0);
+
+		bytesToTruncate = prevFileLen - offset;
+		pg_atomic_sub_fetch_u64(temp_tables_limit_value, bytesToTruncate);
```

Comment:
```suggestion
		pg_atomic_sub_fetch_u64(temp_tables_limit_value, prevFileLen - offset);
```
and variable bytesToTruncate does not need anymore!